### DISCUSSION TITLE
Approaches to detect sequence outliers within taxonomic groups

### DISCUSSION THREAD

**Discussion by jackscanlan** [AVR-biosecurity-bioinformatics/mimir#58] [General]
Three potential approaches, all using the genetic distance from the "central" sequence in a taxonomic group (eg. within a species) to removal sequences above a calculated threshold:
1. Threshold based on genetic distance (eg. 0.05): any sequence >=0.05 distance from the central sequence will be removed
2. Threshold based on a z-score (eg. 3): any sequence with a distance more than 3 standard deviations from the mean will be removed
3. Threshold based on a modified z-score (eg. 3.5): any sequence with a distance more than 3.5 units (?) from the median will be removed
 
Issues with the approaches:
1. Some taxonomic groups may have real intra-group divergence greater than the threshold, which would have the effect of removing real variation from the database and possibly reducing the accuracy of classification
2. Traditional (mean-based) z-score requires a relatively large group size before large outliers are rejected - eg. a 0.1 outlier will only be rejected from a group at threshold 3 once the number of near-zero distance sequences hits 10. This seems too permissive for a "majority rules"-style consensus approach.
3. Modified (median-based) z-score has the unfortunate property of rejecting any sequences with distance >0 if majority of sequences in the group have distance 0 (see: https://stats.stackexchange.com/questions/339932/iglewicz-and-hoaglin-outlier-test-with-modified-z-scores-what-should-i-do-if-t): sequences with any non-zero genetic distance have an infinite/undefined z-score. In this case, the test is too conservative and will remove all intra-group variation if a particular haplotype is over-represented. 

I know you've thought about this before, @alexpiper, but thought it would be good to put it all down somewhere. At the moment just going to implement Approach 1 (flat threshold) with a user-defined distance and can suss out better approaches later.

**Comment by alexpiper** [comment]
Another thing to think about is how we handle 'real' bimodal / multimodal peaks in genetic distance which could occur when a taxonomic 'species' is polyphyletic with the barcode marker. These should probably be retained, but at some point we should validate how this interacts with taxonomic assignment, and whether renaming them (i.e. species_x1, species_x2) improves taxonomic assignment accuracy.

**Reply by jackscanlan** [reply]
That's a good point and something we need to engage with when we validate the pipeline. At the moment the pipeline will just give you a fuzzier output for a given assigned sequence when a taxonomic grouping is polyphyletic (or contains a lot of variation) and I think giving the user control of how much they want to 'assert' additional taxonomic splitting in those cases could be good. In a biosecurity context, where a target group has this issue, I feel like the user would want to validate internal sequences and their own taxonomy before trusting the pipeline to handle it, but I'm sure we'll figure out best approaches when we start to use the pipeline in more applied contexts.
